docs/cpp/source/notes/inference_mode.rst

```diff
@@ -0,0 +1,122 @@
+
+``c10::InferenceMode`` is a new RAII guard analogous to ``NoGradMode``
+to be used when you are certain your operations will have no interactions
+with autograd (e.g. model training). Compared to ``NoGradMode``, code run
+under this mode gets better performance by disabling autograd related work like
+view tracking and version counter bumps. However, tensors created inside
+``c10::InferenceMode`` has more limitation when interacting with autograd system as well.
+
+``InferenceMode`` can be enabled for a given block of code. Inside ``InferenceMode``
+all newly allocated (non-view) tensors are marked as inference tensors. Inference tensors:
+
+- do not have a version counter so an error will be raised if you try to read their version
+  (e.g., because you saved this tensor for backward).
+- are immutable outside ``InferenceMode``. So an error will be raised if you try to:
+  - mutate their data outside InferenceMode.
+  - mutate them into ``requires_grad=True`` outside InferenceMode.
+  To work around you can make a clone to get a normal tensor before mutating.
+
+A non-view tensor is an inference tensor if and only if it was allocated inside ``InferenceMode``.
+A view tensor is an inference tensor if and only if the tensor it is a view of is an inference tensor.
+
+Inside an ``InferenceMode`` block, we make the following performance guarantees:
+
+- Like ``NoGradMode``, all operations do not record ``grad_fn`` even if their ``requires_grad=True``.
+  This applies to both inference tensors and normal tensors.
+- View operations on inference tensors do not do view tracking. View and base inference tensors are
+  indistinguishable.
+- Inplace operations on inference tensors are guaranteed not to do a version bump.
+
+For more implementation details of ``InferenceMode`` please see the `InferenceMode RFC <https://github.com/pytorch/rfcs/pull/17>`_.
+Currently this guard is only available in C++ frontend, adding python frontend support
+is tracked in #56608.
+
+Migration guide from ``AutoNonVariableTypeMode``
+------------------------------------------------
+
+In production use of PyTorch for inference workload, we have seen a proliferation
+of uses of the C++ guard ``AutoNonVariableTypeMode`` (now ``AutoDispatchBelowADInplaceOrView``),
+which disables autograd, view tracking and version counter bumps. Unfortunately,
+current colloquial of this guard for inference workload is unsafe: it's possible to
+use ``AutoNonVariableTypeMode`` to bypass PyTorch's safety checks and result in
+silent wrong results, e.g. PyTorch throws an error when tensors saved for backwards
+are subsequently mutated, but mutation happens inside ``AutoNonVariableTypeMode`` will
+silently bypass the check and returns wrong gradient to users.
+
+When current users of ``AutoNonVariableTypeMode`` think about migrating, the following
+steps might help you decide the best alternatives:
+
+1. Users trying to run workload in inference only mode (like loading a pretrained JIT model and
```

Comment:
Should we recommend loading the model itself in inference mode? So that the loaded weights are directly inference tensors?